File: packages/julia/src/Seguranca/Ataques/AtaqueBytesFixos.jl
"""
Versões antigas dessa cifra tinham um header fixo ou um marcador constante. Esse
ataque gera vários tokens de textos diferentes e procura qualquer posição de
byte que NUNCA muda — uma âncora que um atacante pode usar pra recuperar a chave.
"""
struct AtaqueBytesFixos <: Ataque
    amostras::Int
end

AtaqueBytesFixos(; amostras::Integer = 30) = AtaqueBytesFixos(Int(amostras))

nome(::AtaqueBytesFixos) = "Bytes fixos entre tokens"

function executar(a::AtaqueBytesFixos, alvo::AlvoCriptografico)
    tokens = String[]
    for i in 0:(a.amostras-1)
        letra = string(Char(65 + mod(i, 26)))
        texto = "TEXTO_VARIADO_$i" * repeat(letra, mod(i, 10))
        push!(tokens, encrypt(alvo, texto)[4:end])
    end

    decodificados = [base64url_decode(alvo, t) for t in tokens]
    tamanho_minimo = minimum(length.(decodificados))

    posicoes_fixas = Int[]
    for i in 1:tamanho_minimo
        valores = Set(d[i] for d in decodificados)
        if length(valores) == 1
            push!(posicoes_fixas, i - 1)
        end
    end

    if !isempty(posicoes_fixas)
        return ResultadoAtaque(
            nome(a),
            true,
            "alta",
            "$(length(posicoes_fixas)) posição(ões) de byte fixas em $(a.amostras) tokens: " *
            join(posicoes_fixas[1:min(10, length(posicoes_fixas))], ", "),
            Dict{String,Any}("posicoes" => posicoes_fixas),
        )
    end

    return ResultadoAtaque(
        nome(a),
        false,
        "info",
        "0 de $tamanho_minimo posições fixas em $(a.amostras) tokens",
    )
end

File: packages/dart/lib/Seguranca/Ataques/AtaqueBytesFixos.dart
import 'package:criptografia/Seguranca/AlvoCriptografico.dart';
import 'package:criptografia/Seguranca/Ataque.dart';
import 'package:criptografia/Seguranca/ResultadoAtaque.dart';
import 'package:criptografia/Seguranca/Util.dart';

/// As primeiras versões dessa cifra tinham um header fixo (43 bytes idênticos
/// sempre) ou um marcador constante (o '$' na posição 64). Esse ataque gera
/// vários tokens de textos diferentes e procura qualquer posição de byte que
/// NUNCA muda - isso é uma âncora que um atacante pode usar pra recuperar a
/// chave.
class AtaqueBytesFixos extends Ataque {
  final int _amostras;

  AtaqueBytesFixos({int amostras = 30}) : _amostras = amostras;

  @override
  String nome() => 'Bytes fixos entre tokens';

  @override
  ResultadoAtaque executar(AlvoCriptografico alvo) {
    final tokens = <String>[];
    for (var i = 0; i < _amostras; i++) {
      final texto =
          'TEXTO_VARIADO_$i' +
          repetir(String.fromCharCode(65 + (i % 26)), i % 10);
      tokens.add(alvo.encrypt(texto).substring(alvo.prefixo().length));
    }

    final decodificados = tokens.map(alvo.base64urlDecode).toList();
    final tamanhoMinimo = decodificados
        .map((d) => d.length)
        .reduce((a, b) => a < b ? a : b);

    final posicoesFixas = <int>[];
    for (var i = 0; i < tamanhoMinimo; i++) {
      final valores = decodificados.map((d) => d[i]).toSet();
      if (valores.length == 1) {
        posicoesFixas.add(i);
      }
    }

    if (posicoesFixas.isNotEmpty) {
      return ResultadoAtaque(
        nome(),
        true,
        Severidade.alta,
        '${posicoesFixas.length} posição(ões) de byte fixas em $_amostras '
            'tokens: ${posicoesFixas.take(10).join(', ')}',
        {'posicoes': posicoesFixas},
      );
    }

    return ResultadoAtaque(
      nome(),
      false,
      Severidade.info,
      '0 de $tamanhoMinimo posições fixas em $_amostras tokens',
    );
  }
}
